### ISSUE TITLE
incorrect highlight of command with variable

### ISSUE THREAD

**Issue by msftrncs** [PowerShell/PSReadLine#836]
The syntax highlighting often gives the user the idea of what is going to happen, but in a few situations it seems to indicate the wrong things.

```PowerShell
$a='goodbye'
hello$a # 'hello' is yellow, '$a' is green like its going to be expanded, but it doesn't
# instead, "The term 'hello$a' is not recognized as the name of a …" is returned.
```

I think it is safe to say, everything highlighted yellow, does not allow expansion

```PowerShell
function hello () {echo $args}
$a= @{hello=3;goodbye=4}
$a[hello] # 'hello]' is yellow, like it could be a function call, but function calls seem not to be permitted in array index
$a[hello goodbye] # as you type, it looks like its going to be valid, but the prompt knows, its still red at the end.
$a[(hello goodbye)] # this is accepted
```
![image](https://user-images.githubusercontent.com/26179051/50131055-02d52480-0246-11e9-99c9-391a50365b97.png)

This seems to really confuse the syntax highlighting:
```PowerShell
$a[(hello$b goodbye)]
```
![image](https://user-images.githubusercontent.com/26179051/50131831-abd14e80-0249-11e9-8582-876fff2f0323.png)

Environment data
----------------
```
PS version: 6.1.1
PSReadline version: 2.0.0-beta3
os: 10.0.18298.1000 (WinBuild.160101.0800)
PS file version: 6.1.1
```

**Comment by lzybkr** [comment]
The first case is easy to fix, PSReadLine shouldn't look at nested tokens when the TokenFlag `CommandName` is set on the token.

Your other examples are accurately representing how the PowerShell parser works. When the parser doesn't find an expression for an index, it assumes you're missing a closing `]`. It inserts an artificial `]` and keeps parsing, finding a valid command invocation.

**Comment by msftrncs** [comment]
Quickly realized this occurs in a few other situations as well.  Its like the syntax highlighting is eager to find a command.

(assume 'hello' is the function defined in the earlier post)
```PowerShell
[int]hello 3 # prompt remains red, but 'hello' is yellow
[invalid hello 3 # eager to have a command, 'hello' is yellow
[$b]3 # all the text except $b is yellow, $b is green, but no expansion is allowed here, and the whole thing is actually tokenized as a command name
345+65.com # 65.com becomes yellow, but this is an expression, so commands are no longer possible.
$a = hello 3 # this is valid
$a = 4 + hello 3 # but this is not, yet 'hello' is highlighted yellow
$a = , hello 3 # a command does not seem to work after a comma (its an unary expression operator), yet 'hello' is yellow
```

In the cases were a command is indicated that shouldn't be, there seems to be no treatment for the explicit case of being in expression mode.  It would appear that the syntax highlighter is confident, that the first thing on a command line that is not 'something else' must be a command, where, in reality, the only thing that can be a command, is the first token, and only if its not valid as an expression first.

And so that last statement is not misinterpreted, I understand that expressions, sub-expressions, terminators, and pipelines all return back to argument mode, thus another command could appear on the same statement line; I am only showing statements on a new line for simplicity.

**Comment by msftrncs** [comment]
For follow up on this, its not possible to prevent the nested tokens using the token flag `CommandName` because its also set on the expandable argument token following the `&` and `.` operators.   So it would seem to me that the only changes PSReadLine can make is to handle parameter tokens that have the `CommandName` flag set.

**Comment by msftrncs** [comment]
I almost missed my last comment from Feb 2, the last line of code, showing the command name highlight bleeding the rest of the statement:

![image](https://user-images.githubusercontent.com/26179051/62408704-738d3f00-b592-11e9-98b1-e18c0f9990ba.png)
(image might be from 2.0.0beta1)

At the time, I only entered the comment part here in the markdown, and not on the command line.  Revisiting it made the issue that much larger when the comment was included.

**Comment by msftrncs** [comment]
Also see https://github.com/PowerShell/PowerShell/issues/10275.

**Comment by msftrncs** [comment]
Closing, as the items PSReadLine is responsible for have been fixed in #989.
